File: mkr-android-lib/src/main/java/com/mkrworld/androidlib/ui/adapter/BaseViewHolder.java
package com.mkrworld.androidlib.ui.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;

/**
 * Created by himanshu on 08/09/15.
 */
public abstract class BaseViewHolder<V> extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

    protected View mParent;
    private VHClickable mClickCallback;
    private VHLongClickable mLongClickCallback;
    private AdapterItem mAdapterItem;

    /**
     * Constructor
     *
     * @param itemView
     */
    public BaseViewHolder(View itemView) {
        super(itemView);
        mParent = itemView;
    }

    public View getParent() {
        return mParent;
    }

    public Context getContext() {
        return mParent.getContext();
    }

    public void bindFeedItem(AdapterItem<V> adapterItem) {
        mAdapterItem = adapterItem;
        bindData(adapterItem.getBindingData());
    }

    /**
     * Method called, whenever view bind with Recycler View
     * @param v
     */
    protected abstract void bindData(V v);

    public AdapterItem getAdapterItem() {
        return mAdapterItem;
    }

    @Override
    public void onClick(View v) {
        if (mClickCallback != null) mClickCallback.onViewHolderClicked(this, v);
    }

    @Override
    public boolean onLongClick(View v) {
        if (mLongClickCallback != null) {
            mLongClickCallback.onViewHolderLongClicked(this, v);
            return true;
        }
        return false;
    }

    /**
     * Method to set the VHClickCallback
     *
     * @param callback
     */
    public void setVHClickCallback(VHClickable callback) {
        mClickCallback = callback;
    }

    /**
     * Method to set the VHLongClickCallback
     *
     * @param callback
     */
    public void setVHLongClickCallback(VHLongClickable callback) {
        mLongClickCallback = callback;
    }

    /**
     * Callback to notifyTaskResponse that user clicked the viewItem
     */
    public interface VHClickable {
        /**
         * Method called when user click NewsAdapterItemHandler view in adapter Item
         *
         * @param holder
         * @param view
         */
        void onViewHolderClicked(BaseViewHolder holder, View view);
    }

    /**
     * Callback to notifyTaskResponse that user long-clicked the viewItem
     */
    public interface VHLongClickable {
        /**
         * Method called when user long-click NewsAdapterItemHandler view in adapter Item
         *
         * @param holder
         * @param view
         */
        void onViewHolderLongClicked(BaseViewHolder holder, View view);
    }
}

File: mkr-android-lib/src/main/java/com/mkrworld/androidlib/ui/adapter/BaseViewHolder.kt
package com.mkrworld.androidlib.ui.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import javax.xml.transform.Source

/**
 * Created by mkr on 3/4/18.
 */
abstract class BaseViewHolder<MKR> : RecyclerView.ViewHolder, View.OnClickListener, View.OnLongClickListener {
    protected var mParent : View? = null
    private var mClickCallback : VHClickable? = null
    private var mLongClickCallback : VHLongClickable? = null
    private var mBaseAdapterItem : BaseAdapterItem<*>? = null

    /**
     * Constructor
     *
     * @param itemView
     */
    constructor(itemView : View) : super(itemView) {
        mParent = itemView
    }

    fun getParent() : View {
        return mParent !!
    }

    fun getContext() : Context {
        return mParent !!.getContext()
    }

    fun bindFeedItem(baseAdapterItem : BaseAdapterItem<*>) {
        mBaseAdapterItem = baseAdapterItem
        bindData((baseAdapterItem as BaseAdapterItem<MKR>).getData())
    }

    /**
     * Method called, whenever view bind with Recycler View
     * @param mkr
     */
    protected abstract fun bindData(mkr : MKR)

    fun getAdapterItem() : BaseAdapterItem<*> {
        return mBaseAdapterItem !!
    }

    override fun onClick(v : View) {
        mClickCallback?.onViewHolderClicked(this, v)
    }

    override fun onLongClick(v : View) : Boolean {
        if (mLongClickCallback != null) {
            mLongClickCallback !!.onViewHolderLongClicked(this, v)
            return true
        }
        return false
    }

    /**
     * Method to set the VHClickCallback
     *
     * @param callback
     */
    fun setVHClickCallback(callback : VHClickable) {
        mClickCallback = callback
    }

    /**
     * Method to set the VHLongClickCallback
     *
     * @param callback
     */
    fun setVHLongClickCallback(callback : VHLongClickable) {
        mLongClickCallback = callback
    }


    /**
     * Callback to notifyTaskResponse that user clicked the viewItem
     */
    interface VHClickable {
        /**
         * Method called when user click NewsAdapterItemHandler view in adapter Item
         *
         * @param holder
         * @param view
         */
        fun onViewHolderClicked(holder : BaseViewHolder<*>, view : View)
    }

    /**
     * Callback to notifyTaskResponse that user long-clicked the viewItem
     */
    interface VHLongClickable {
        /**
         * Method called when user long-click NewsAdapterItemHandler view in adapter Item
         *
         * @param holder
         * @param view
         */
        fun onViewHolderLongClicked(holder : BaseViewHolder<*>, view : View)
    }
}